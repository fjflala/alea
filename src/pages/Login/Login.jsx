/**
 * Module dependencies
 */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

/**
 * Actions
 */
import { authLogIn } from '../../modules/Authentication/actions';

/**
 * UI Components
 */
import Button from '../../components/Button';
import Card from '../../components/Card';
import Input from '../../components/Input';

/**
 * Import styles
 */
import './styles.scss';

/**
 * Represents a view of Login
 */
export const Login = ({
  history,
  location,
  isAuthenticated,
  token,
  error,
  logIn,
}) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState(null);

  const onSubmit = (event) => {
    event.preventDefault();

    logIn({
      email,
      password,
    });
  };

  useEffect(() => {
    if (isAuthenticated && token.length > 0) {
      const nextPage = (location.state && location.state.from && location.state.from.pathname) || '/users';
      history.push(nextPage);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [token, isAuthenticated]);

  return (
    <section className="login">
      <Card className="login__card">
        <img
          src="https://www.alea.com/wp-content/themes/alea/library/images/logo-alea.svg"
          alt="Alea"
          className="login__alea-logo"
        />
        {error.length > 0 && (
          <p className="login__error-message">
            The email or password is not working.
            <br />
            You can test with the mocked values
          </p>
        )}
        <form onSubmit={onSubmit} className="login__form">
          <Input type="text" placeholder="Email" onChange={e => setEmail(e.target.value)} />
          <Input type="password" placeholder="Password" onChange={e => setPassword(e.target.value)} />
          <Button type="submit" modifier="primary" className="login__button">Login</Button>
        </form>
        <p>
          Or signup, click &nbsp;
          <Link to="/signup">here!</Link>
        </p>
      </Card>
      <Card className="login__card">
        <h4>Mocked Values</h4>
        <strong>User:</strong>
        &nbsp;eve.holt@reqres.in
        <br />
        <strong>Password:</strong>
        &nbsp;cityslicka
      </Card>
    </section>
  );
};

/**
 * PropTypes definitions 
 */
Login.propTypes = {
  history: PropTypes.shape({}).isRequired,
  location: PropTypes.shape({}).isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  token: PropTypes.string,
  error: PropTypes.string,
  logIn: PropTypes.func.isRequired,
};

Login.defaultProps = {
  token: '',
  error: '',
};

/**
 * Connect to Store
 */
const mapStateToProps = state => ({
  isAuthenticated: state.authentication.isAuthenticated,
  token: state.authentication.token,
  error: state.authentication.error,
});

const mapDispatchToProps = { logIn: authLogIn }

/**
 * Expose component
 */
export default connect(mapStateToProps, mapDispatchToProps)(Login);
