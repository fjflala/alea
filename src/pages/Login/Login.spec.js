import React from 'react';
import renderer from 'react-test-renderer';
import { Login } from './Login';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { mount } from 'enzyme';

describe('<Login />', () => {
  let props; 
  
  beforeEach(() => {
    props = {
      isAuthenticated: false,
      token: '',
      error: '',
      logIn: jest.fn(),
      history: {},
      error: '',
      location: {},
    };
  });

  it('should correctly render', () => {
    const tree = renderer
      .create(<Router history={createMemoryHistory({route: ['/login']})}>
        <Login {...props} />
      </Router>)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('should call register method when click on register button', () => {
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/signup']})}>
      <Login {...props} />
    </Router>);
    
    wrapper.find('form').simulate('submit');
    
    expect(props.logIn).toHaveBeenCalled();
  });

  it('should show an error message', () => {
    props.error = 'Error while trying to log the user into the application';
    const wrapper = mount(<Router history={createMemoryHistory({route: ['/signup']})}>
      <Login {...props} />
    </Router>);
    
    expect(wrapper.find('.login__error-message').length).toBe(1);
  });
});