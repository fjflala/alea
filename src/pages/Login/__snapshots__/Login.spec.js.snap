// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should correctly render 1`] = `
<section
  className="login"
>
  <div
    className="ui-card login__card"
  >
    <img
      alt="Alea"
      className="login__alea-logo"
      src="https://www.alea.com/wp-content/themes/alea/library/images/logo-alea.svg"
    />
    <form
      className="login__form"
      onSubmit={[Function]}
    >
      <input
        className="ui-input "
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Email"
        type="text"
      />
      <input
        className="ui-input "
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        type="password"
      />
      <button
        className="ui-button ui-button--primary login__button"
        onClick={[Function]}
        type="submit"
      >
        Login
      </button>
    </form>
    <p>
      Or signup, click  
      <a
        href="/signup"
        onClick={[Function]}
      >
        here!
      </a>
    </p>
  </div>
  <div
    className="ui-card login__card"
  >
    <h4>
      Mocked Values
    </h4>
    <strong>
      User:
    </strong>
     eve.holt@reqres.in
    <br />
    <strong>
      Password:
    </strong>
     cityslicka
  </div>
</section>
`;
